### PR TITLE
Implement the search function

### PR DESCRIPTION
**Pull request by changhiskhan**
note that this is pulling into the ann branch

### CODE REVIEW

File: rust/src/index/ann/distance.rs [hunk 1]
```diff
@@ -6,29 +6,42 @@
...
-            sub = multiply(&sub, &sub).unwrap();
-            sum(&sub).unwrap().sqrt()
+            euclidean_distanc(from, arr).unwrap()  // how to handle NAs ?
         })
         .collect();
     Ok(scores)
 }
 
+
+pub fn euclidean_distanc(
```
Comment (eddyxu) [inline rust/src/index/ann/distance.rs:36]: `distanc` misspell?

File: rust/src/index/ann/sort.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            return Ok(());
+        } else if left > topk {
+            high = left;
+        } else if left < topk {
+            low = left;
+        }
+    }
+}
+
+/// Returns the number of elements that are smaller than the pivot value
```
Comment (eddyxu) [inline rust/src/index/ann/sort.rs:31]: This is fancy stuff.  isnt there some partition alrogithm / crates out there?

Comment (changhiskhan) [reply rust/src/index/ann/sort.rs:34]: rayon has something but the partitioning logic isn't a public function. I'd rather not maintain our own so if you see/know a good one lmk.

**Comment by changhiskhan** [comment]
actually this is not ready yet. there are some edge cases where it just loops forever. i need to investigate some more

File: rust/src/index/ann/sort.rs (hunk 2 above)
Comment (changhiskhan) [reply rust/src/index/ann/sort.rs:34]: also the other problem is it needs to "co-partition" the scores and row_id's together. So really we're looking for arg-topk

**Pull request closed without merging**